controllers/CartController.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace app\controllers;
+
+use app\core\Controller;
+use app\core\Request;
+use app\core\Response;
+use app\models\CartItem;
+
+class CartController extends Controller
+{
+    public function cart()
+    {
+        return $this->render("cart");
+    }
+
+    public function addToCart(Request $request, Response $response)
+    {
+        $cartItem = new CartItem();
+
+        if ($request->isPost()) {
+            $cartItem->loadData($request->getBody());
+            $cartItem->setCartID($_SESSION["cartID"]);
+
+            $sql = [
+                "cart_id" => $cartItem->cart_id,
+                "product_id" => $cartItem->product_id
+            ];
+
+            $itemAdded = $cartItem->findOne($sql);
+
+            if (!$itemAdded) {
+                $quantity = 1;
+                $cartItem->setQuantity($quantity);
+
+                if ($cartItem->validate() && $cartItem->save()) {
+                    $response->redirect("/cart");
+                }
+            } else {
+                if ($cartItem->quantity == 0) {
+                    $newQuantity = $itemAdded->quantity + 1;
+                    $cartItem->update(["quantity" => $newQuantity], $sql);
+                    $response->redirect("/cart");
+                }
+            }
+        } else {
+            $cartItems = $cartItem->getByID(["cart_id" => $_SESSION["cartID"]]);
+            $params = [
+                "cartItems" => $cartItems
+            ];
+            return $this->render("cart", $params);
+        }
+    }
+
+    public function getCartItems(Request $request, Response $response)
+    {
+        $cartItem = new CartItem();
+
+        $items = $cartItem->getByID(["cart_id" => $_SESSION["cartID"]]);
+        $items = json_encode($items);
+
+        return $items;
```

Comment:
Oh I was intending to make this return json data whenever there's an API call. The others were coded with Server-Side Rendering in mind. Therefore, they maybe (or are) unclear at the moment since I'm switching the server to a more API friendly and leave to frontend to frontent.

Reply:
That makes sense. It's hard to distinguish if something is an API versus rendering. Maybe a naming convention. Another way to think about this is to have a set of "output" classes. For example, you would have a `JsonOutput` and a `HtmlOutput` class. Then, in the control plane that calls into these methods, it would make a distinction on how to render it back. Something like:

```php
abstract class BaseOutput {
  abstract public function render();
}

class JsonOutput extends BaseOutput {
  public function __construct($jsonObject) {
    $this->data = $jsonObject;
  }

  public function setData($jsonObject) {
    $this->data = $jsonObject;
  }

  public function render() {
    return json_encode($this->data);
  }
}

class HtmlOutput extends BaseOutput {
  public function __construct() {
    $this->combinedHtml = '';
  }

  public function append($html) {
    $this->combinedHtml .= $html;

  public function render() {
    return $this->combinedHtml;
  }
}
```

The change in your Application.php would be something like:

```php
$output = $this->router->resolve();
echo $output->render();
```

```php
public function apiGetCart() {
  $cartContent = ...;
  return new JsonOutput($cartContent);
}

public function renderCart() {
  $html = new HtmlOutput();
  $html.append('<pre>');
  $html.append('Something</pre>');
  return $html;
}
```

Reply:
If this becomes a strictly API and no HTML rendering, are you planning on setting up something like React or Vue?

Reply:
I am not really familiar with React or Vue at the moment. But yes I'm planning to switch to them in the future. Back then, I was embedding php code with html. Now I want to separate the backend and the frontend so as to be more flexible.